tls/s2n_x509_validator.c

```diff
@@ -0,0 +1,446 @@
+#endif /* defined(OPENSSL_IS_BORINGSSL) && !defined(OCSP_RESPONSE_STATUS_SUCCESSFUL) */
+
+/* our friends at openssl love to make backwards incompatible changes */
+#if !defined(LIBRESSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER >= 0x10100000L
+#define OCSP_GET_CERTS(a) OCSP_resp_get0_certs(a)
+#else
+#define OCSP_GET_CERTS(a) a->certs
+#endif
+
+uint8_t s2n_x509_ocsp_stapling_supported(void) {
+    return S2N_OCSP_STAPLING_SUPPORTED;
+}
+
+void s2n_x509_trust_store_init(struct s2n_x509_trust_store *store) {
+    store->trust_store = NULL;
+}
+
+uint8_t s2n_x509_trust_store_has_certs(struct s2n_x509_trust_store *store) {
+    return store->trust_store ? (uint8_t) 1 : (uint8_t) 0;
+}
+
+int s2n_x509_trust_store_from_system_defaults(struct s2n_x509_trust_store *store) {
+    if(!store->trust_store) {
+        store->trust_store = X509_STORE_new();
+    }
+
+    int err_code = X509_STORE_set_default_paths(store->trust_store);
+
+    if (!err_code) {
+        s2n_x509_trust_store_wipe(store);
+        return -1;
+    }
+
+    X509_STORE_set_flags(store->trust_store, X509_VP_FLAG_DEFAULT);
+
+    return 0;
+}
+
+int s2n_x509_trust_store_from_ca_file(struct s2n_x509_trust_store *store, const char *ca_file, const char *path) {
+
+    if(!store->trust_store) {
+        store->trust_store = X509_STORE_new();
+    }
+
+    int err_code = X509_STORE_load_locations(store->trust_store, ca_file, path);
+
+    if (!err_code) {
+        s2n_x509_trust_store_wipe(store);
+        return -1;
+    }
+
+    unsigned long flags = X509_VP_FLAG_DEFAULT;
+
+#if defined(X509_V_FLAG_PARTIAL_CHAIN)
+    flags |=  X509_V_FLAG_PARTIAL_CHAIN;
+#else
+    flags |= 0x80000;
+#endif
+
+    X509_STORE_set_flags(store->trust_store, flags);
+
+    return 0;
+}
+
+void s2n_x509_trust_store_wipe(struct s2n_x509_trust_store *store) {
+    if (store->trust_store) {
+        X509_STORE_free(store->trust_store);
+        store->trust_store = NULL;
+    }
+}
+
+int s2n_x509_validator_init_no_checks(struct s2n_x509_validator *validator) {
+    validator->trust_store = NULL;
+    validator->cert_chain = NULL;
+    validator->validate_certificates = 0;
+    validator->check_stapled_ocsp = 0;
+
+    return 0;
+}
+
+int s2n_x509_validator_init(struct s2n_x509_validator *validator, struct s2n_x509_trust_store *trust_store, uint8_t check_ocsp) {
+    notnull_check(trust_store);
+    validator->trust_store = trust_store;
+
+    validator->validate_certificates = 1;
+    validator->check_stapled_ocsp = check_ocsp;
+
+    validator->cert_chain = NULL;
+    if (validator->trust_store->trust_store) {
+        validator->cert_chain = sk_X509_new_null();
+    }
+
+    return 0;
+}
+
+void s2n_x509_validator_wipe(struct s2n_x509_validator *validator) {
+    if (validator->cert_chain) {
+        sk_X509_pop_free(validator->cert_chain, X509_free);
+        validator->cert_chain = NULL;
+    }
+
+    validator->trust_store = NULL;
+    validator->validate_certificates = 0;
```

Comment:
I think wiping a validator should leave it in a secure state that attempts to validate certificates (even if it has an empty trust store). The only way to get an insecure validator should be through an explicit API call that does exactly that.

Otherwise calling `s2n_x509_validator_validate_cert_chain` with a `s2n_x509_validator` that's been wiped or hasn't been initialized may silently skip x509 checks rather than failing.

Reply:
good call.